docs/design/pod_evication_and_replacement.md

```diff
@@ -0,0 +1,123 @@
+
+## NoExecute Tolerations
+
+NoExecute tolerations are used to control the behavior of Kubernetes (wrt. to a Pod) when the node
+that the pod is running on is no longer reachable or becomes not-ready.
+
+See the applicable [Kubernetes documentation](https://kubernetes.io/docs/concepts/configuration/taint-and-toleration/) for more info.
+
+## Rules
+
+The rules for eviction & replacement are specified per type of pod.
+
+### Image ID Pods
+
+The Image ID pods are starter to fetch the ArangoDB version of a specific
+ArangoDB image and fetch the docker sha256 of that image.
+They have no persistent state.
+
+- Image ID pods can always be evicted from any node
+- Image ID pods can always be restarted on a different node.
+  There is no need to replace an image ID pod.
+- `node.kubernetes.io/unreachable:NoExecute` toleration time is set very low (5sec)
+- `node.kubernetes.io/not-ready:NoExecute` toleration time is set very low (5sec)
+
+### Coordinator Pods
+
+Coordinator pods run an ArangoDB coordinator as part of an ArangoDB cluster.
+They have no persistent state, but do have a unique ID.
+
+- Coordinator pods can always be evicted from any node
+- Coordinator pods can always be replaced with another coordinator pod with a different ID on a different node
+- `node.kubernetes.io/unreachable:NoExecute` toleration time is set low (15sec)
+- `node.kubernetes.io/not-ready:NoExecute` toleration time is set low (15sec)
+
+### DBServer Pods
+
+DBServer pods run an ArangoDB dbserver as part of an ArangoDB cluster.
+It has persistent state potentially tight to the node it runs on and it has a unique ID.
+
+- DBServer pods can be evicted from any node as soon as:
+  - It has been completely drained AND
+  - It is no longer the shard master for any shard
+- DBServer pods can be replaced with another dbserver pod with a different ID on a different node when:
+  - It is not the shard master for any shard OR
+  - For every shard it is the master for, there is an in-sync follower
+- `node.kubernetes.io/unreachable:NoExecute` toleration time is set high to "wait it out a while" (5min)
+- `node.kubernetes.io/not-ready:NoExecute` toleration time is set high to "wait it out a while" (5min)
+
+### Agent Pods
+
+Agent pods run an ArangoDB dbserver as part of an ArangoDB agency.
+It has persistent state potentially tight to the node it runs on and it has a unique ID.
+
+- Agent pods can be evicted from any node as soon as:
+  - It is no longer the agency leader AND
+  - There is at least an agency leader that is responding AND
+  - There is at least an agency follower that is responding
+- Agent pods can be replaced with another agent pod with the same ID but whiped persistent state on a different node when:
+  - The old pod is known to be deleted (e.g. explicit eviction)
+- `node.kubernetes.io/unreachable:NoExecute` toleration time is not set to "wait it out forever"
+- `node.kubernetes.io/not-ready:NoExecute` toleration time is not set "wait it out forever"
```

Comment:
Contrary to what we have discussed last week, I fear I have second thoughts here. I still have to check if the wiping of the data actually makes the leader resend and resynchronize all data. Will do that this afternoon. But I think we should make it so that this formulation is OK.